docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,128 @@
+
+```
+call RuntimeHelpers.CheckConstraintsForMethodCallThatFollows
+br.false ....
+push arg0
+push arg1
+...
+push argN
+call CodeWithMoreStringentConstraints
+```
+
+The sequence must be within the same basic block and there must not be any other call/callvirt between `CheckConstraintsForMethodCallThatFollows` and the more constrained method call.
+
+NOTE: This would only check constraints that the runtime recognizes. It would still be responsibility of the language compiler to add checks for constraints the runtime doesn't recognize, such as `unmanaged` in C#.
+
+The proposal is to:
+
+* Stop doing the verification-level constraint validation when compiling `call` prefixed with `CheckConstraintsForMethodCallThatFollows` (i.e. the check on uninstantiated forms).
+* Keep constraint validation of the instantiated form. Components operating on instantiated method bodies (such as code generators or debuggers) would do the necessary macro expansion of the _constraints bridge_ IL sequence:
+    * If the constraints are not met, treat the sequence as an unconditional jump to the branch that handles false return from `CheckConstraintsForMethodCallThatFollows`.
+    * If the constraints are met, treat the sequence as a normal call.
+    * If the constraints may be met (e.g. due to shared code), generate a runtime check.
+
+### Alternative option: allow loading constraint-failing MethodTables
+
+Limiting the constraint-violating code to `call` instruction is required to simplify code generation. Allowing `MethodTable`s that don't satisfy their own constraints to load sounds like trouble. We could potentially allow them to load and then try to prevent them from leaking out. Calling this out for completeness, I don't know if it's a good idea.
+
+### Potential shortcut: Stop validating constraints
+
+One more avenue could be to drop all constraint checks within the runtime and make constraints source compiler's concern only. We already have a constraint (`unmanaged` in C#) that is only known to the source compiler and not enforced at run time. This would likely require some thoughts around failure modes when invalid IL within a method body IL is encountered. This document is not going to explore this direction further since the assumption is that we do not want to drop constraint checks from the runtime, we just want to delay them.
+
+## API proposal
+
+Runtime support for this will be indicated by following property:
+
+```diff
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeHelpers
+    {
++        public static bool CheckConstraintsForMethodCallThatFollows();
+    }
+}
+
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeFeature
+    {
++        /// <summary>
++        /// Represents a runtime feature where constraint validation can be delayed until run time.
++        /// </summary>
++        public const string DelayedConstraintCheck = nameof(DelayedConstraintCheck);
```

Comment:
Nit: This is not strictly required. The existence of this feature can be inferred from existence of the `CheckConstraintsForMethodCallThatFollows` API.